acados/ocp_nlp/ocp_nlp_sqp.c

```diff
@@ -785,76 +701,271 @@ int ocp_nlp_sqp(void *config_, void *dims_, void *nlp_in_, void *nlp_out_,
             if (opts->print_level > 1)
             {
                 printf("\n Failed to solve the following QP:\n");
-                if (opts->print_level > sqp_iter + 1)
-                    print_ocp_qp_in(nlp_mem->qp_in);
+                if (opts->print_level)
+                    print_ocp_qp_in(qp_in);
             }
 
             mem->status = ACADOS_QP_FAILURE;
+            mem->sqp_iter = sqp_iter;
+            mem->time_tot = acados_toc(&timer0);
+
             return mem->status;
         }
 
-        // globalization
+        /* globalization */
+        // NOTE on timings: currently all within globalization is accounted for within time_glob.
+        //   QP solver times could be also attributed there alternatively. Cleanest would be to save them seperately.
         acados_tic(&timer1);
-        double alpha = ocp_nlp_line_search(config, dims, nlp_in, nlp_out, nlp_opts, nlp_mem, nlp_work);
+        bool do_line_search = true;
+        if (opts->nlp_opts->glob_SOC && opts->nlp_opts->globalization == MERIT_BACKTRACKING)
+        {
+            // NOTE: following Waechter2006:
+            // Do SOC
+            // 1.if "the first trial step size alpha_k,0 has been rejected and
+            // 2. if the infeasibility would have increased when accepting the previous step
+
+            // preliminary line search
+            alpha = ocp_nlp_line_search(config, dims, nlp_in, nlp_out, nlp_opts, nlp_mem, nlp_work, 1);
+
+            if (alpha < 1.0)
+            {
+                // Second Order Correction (SOC): following Nocedal2006: p.557, eq. (18.51) -- (18.56)
```

Comment:
It looks like the page and equations are different in my version of the book, can you maybe write the chapter number and title too?